File: src/main/java/io/eliez/fintools/bin2brand/PrefixClassifier.java
package io.eliez.fintools.bin2brand;

import one.util.streamex.EntryStream;
import one.util.streamex.StreamEx;

import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public class PrefixClassifier {

    private final List<LabeledPrefixRange> prefixes;

    public PrefixClassifier(Map<String, List<PrefixRange>> rangesTable) {
        this.prefixes = loadPrefixes(rangesTable);
    }

    public Optional<String> findLabel(String target) {
        return StreamEx.of(prefixes)
                .findFirst(prefix -> prefix.contains(target))   //NOSONAR -- likely a Sonarqube error
                .map(LabeledPrefixRange::label);
    }

    public static PrefixClassifier fromTextTable(Map<String, String> rangesTextTable, RangePattern pattern) {
        return new PrefixClassifier(EntryStream.of(rangesTextTable)
                .mapValues(vv -> StreamEx.split(vv, ",").map(v -> PrefixRangeParser.parse(v, pattern)).toList())
                .toMap());
    }

    private List<LabeledPrefixRange> loadPrefixes(Map<String, List<PrefixRange>> rangesTable) {
        final List<LabeledPrefixRange> ranges = EntryStream.of(rangesTable)
                .flatMapKeyValue((k, vv) -> StreamEx.of(vv).map(v -> new LabeledPrefixRange(k, v)))
                .sorted(Comparator.comparingInt(LabeledPrefixRange::length)
                        .reversed()
                        .thenComparing(LabeledPrefixRange::lower))
                .toList();
        validateRanges(ranges);
        return ranges;
    }

    private void validateRanges(List<LabeledPrefixRange> ranges) {
        StreamEx.of(ranges)
                .forPairs((npr1, npr2) -> {
                    if ((!(npr1.label().equals(npr2.label()))
                            && npr1.length() == npr2.length())
                            && (npr1.upper().compareTo(npr2.lower()) >= 0)) {
                        throw new ConflictingRangesException(
                                String.format("illegal intersection between ranges '%s' and '%s'",
                                        npr1.toShortString(), npr2.toShortString()));
                    }
                });
    }
}

File: src/main/kotlin/io/eliez/fintools/bin2brand/PrefixClassifier.kt
package io.eliez.fintools.bin2brand

import one.util.streamex.StreamEx
import kotlin.comparisons.compareByDescending
import kotlin.comparisons.thenBy

internal class PrefixClassifier(rangesTable: Map<String, List<PrefixRange>>) {

    private val prefixes: List<LabeledPrefixRange>

    init {
        this.prefixes = loadPrefixes(rangesTable)
    }

    fun findLabel(target: String): String? {
        return prefixes.find { it -> it.contains(target) }?.label
    }

    private fun loadPrefixes(rangesTable: Map<String, List<PrefixRange>>): List<LabeledPrefixRange> {
        val ranges = rangesTable.flatMap { entry -> entry.value.map { it -> LabeledPrefixRange(entry.key, it) } }
                .sortedWith(compareByDescending<LabeledPrefixRange> { it.length() }.thenBy { it.lower })
        validateRanges(ranges)
        return ranges
    }

    private fun validateRanges(ranges: List<LabeledPrefixRange>) {
        StreamEx.of(ranges)
                .forPairs { npr1, npr2 ->
                    if (npr1.label != npr2.label &&
                            npr1.length() == npr2.length() &&
                            npr1.upper >= npr2.lower) {
                        throw ConflictingRangesException(
                                String.format("illegal intersection between ranges '%s' and '%s'",
                                        npr1.toShortString(), npr2.toShortString()))
                    }
                }
    }

    companion object {

        @JvmStatic
        fun fromTextTable(rangesTextTable: Map<String, String>, pattern: RangePattern): PrefixClassifier {
            return PrefixClassifier(rangesTextTable.mapValues {
                entry ->
                entry.value.split(",").map { it -> PrefixRangeParser.parse(it, pattern) }
            })
        }
    }
}
